library/objective-c/EnvoyHttpStream.m

```diff
@@ -0,0 +1,250 @@
+  return ret;
+}
+
+static NSData * to_ios_data(envoy_data data) {
+  // TODO: investigate buffer ownership
+  // Possibly extend/subclass NSData to call envoy_data.release on dealloc and have release drain the underlying Envoy buffer instance
+  return [NSData dataWithBytes:(void *)data.bytes length:data.length];
+}
+
+static EnvoyHeaders * to_ios_headers(envoy_headers headers) {
+  NSMutableDictionary *headerDict = [NSMutableDictionary new];
+  for (uint_fast32_t i = 0; i < headers.length; i++) {
+    envoy_header header = headers.headers[i];
+    NSString *headerKey = [[NSString alloc] initWithBytes:header.key.bytes
+                                                   length:header.key.length
+                                                 encoding:NSUTF8StringEncoding];
+    NSString *headerValue = [[NSString alloc] initWithBytes:header.value.bytes
+                                                     length:header.value.length
+                                                   encoding:NSUTF8StringEncoding];
+    NSMutableArray *headerValueList = headerDict[headerKey];
+    if (headerValueList == nil) {
+      headerValueList = [NSMutableArray new];
+      headerDict[headerKey] = headerValueList;
+    }
+    [headerValueList addObject:headerValue];
+  }
+  return headerDict;
+}
+
+#pragma mark - c callbacks
+static void ios_on_headers(envoy_headers headers, bool end_stream, void* context) {
+  ios_context *c = (ios_context *)context;
+  EnvoyObserver *observer = c->observer;
+  dispatch_async(observer.dispatchQueue, ^{
+    if (atomic_load(c->canceled)) {
+      return;
+    }
+    observer.onHeaders(to_ios_headers(headers), end_stream);
+  });
+}
+
+static void ios_on_data(envoy_data data, bool end_stream, void* context) {
+  ios_context *c = (ios_context *)context;
+  EnvoyObserver *observer = c->observer;
+  dispatch_async(observer.dispatchQueue, ^{
+    if (atomic_load(c->canceled)) {
+      return;
+    }
+    // TODO: retain data
+    observer.onData(to_ios_data(data), end_stream);
+  });
+}
+
+static void ios_on_metadata(envoy_headers metadata, void* context) {
+  ios_context *c = (ios_context *)context;
+  EnvoyObserver *observer = c->observer;
+  dispatch_async(observer.dispatchQueue, ^{
+    if (atomic_load(c->canceled)) {
+      return;
+    }
+    observer.onMetadata(to_ios_headers(metadata));
+  });
+}
+
+static void ios_on_trailers(envoy_headers trailers, void *context) {
+  ios_context *c = (ios_context *)context;
+  EnvoyObserver *observer = c->observer;
+  dispatch_async(observer.dispatchQueue, ^{
+    if (atomic_load(c->canceled)) {
+      return;
+    }
+    observer.onTrailers(to_ios_headers(trailers));
+  });
+}
+
+static void ios_on_complete(void *context) {
+  ios_context *c = (ios_context *)context;
+  EnvoyObserver *observer = c->observer;
+  dispatch_async(observer.dispatchQueue, ^{
+    // TODO: release stream
+    if (atomic_load(c->canceled)) {
+      return;
+    }
+  });
```

Comment:
Shouldn't this be calling back up to the platform?

Reply:
Yep, some of this is more of a sketch; working headers is the main goal for this PR (with the framework in place to get the rest working).